river/tree/hoeffding_adaptive_tree_classifier.py

```diff
@@ -174,68 +177,50 @@ def learn_one(self, x, y, *, sample_weight=1.0):
-                # enough to be used to predict, so skip it
-                if fn.parent_branch != -999:
-                    leaf_node = fn.node
-                    if leaf_node is None:
-                        leaf_node = fn.parent
-                    dist = leaf_node.leaf_prediction(x, tree=self)
-                    # Option Tree prediction (of sorts): combine the response of all leaves reached
-                    # by the instance
-                    proba = add_dict_values(proba, dist, inplace=True)
+        if self._root is not None:
+            found_nodes = [self._root]
+            if isinstance(self._root, HTBranch):
+                found_nodes = self._root.traverse(x, until_leaf=True)
+            for leaf in found_nodes:
+                dist = leaf.prediction(x, tree=self)
+                # Option Tree prediction (of sorts): combine the response of all leaves reached
+                # by the instance
+                proba = add_dict_values(proba, dist, inplace=True)
             proba = normalize_values_in_dict(proba)
 
         return proba
 
-    def _filter_instance_to_leaves(self, x, split_parent, parent_branch):
-        nodes = []
-        self._tree_root.filter_instance_to_leaves(x, split_parent, parent_branch, nodes)
-        return nodes
+    def _new_leaf(self, initial_stats=None, parent=None):
+        if initial_stats is None:
+            initial_stats = {}
 
-    def _new_learning_node(self, initial_stats=None, parent=None):
         if parent is not None:
             depth = parent.depth + 1
         else:
             depth = 0
 
-        return AdaLearningNodeClassifier(
+        return AdaLeafClassifier(
             stats=initial_stats,
             depth=depth,
             splitter=self.splitter,
             adwin_delta=self.adwin_confidence,
             seed=self.seed,
         )
 
-    def _new_split_node(self, split_test, target_stats=None, depth=0, **kwargs):
-        return AdaSplitNodeClassifier(
-            split_test=split_test,
-            stats=target_stats,
-            depth=depth,
-            adwin_delta=self.adwin_confidence,
-            seed=self.seed,
-        )
-
-    # Override river.tree.BaseHoeffdingTree to include alternate trees
-    def __find_learning_nodes(self, node, parent, parent_branch, found):
-        if node is not None:
-            if isinstance(node, LearningNode):
-                found.append(FoundNode(node, parent, parent_branch))
-            if isinstance(node, SplitNode):
-                split_node = node
-
-                for i in range(split_node.n_children):
-                    self.__find_learning_nodes(
-                        split_node.get_child(i), split_node, i, found
-                    )
-                if split_node._alternate_tree is not None:  # noqa
-                    self.__find_learning_nodes(
-                        split_node._alternate_tree, split_node, -999, found  # noqa
-                    )
+    def _branch_selector(
```

Comment:
Due to the new base classes, this function does not actually build the branch/split node. It only selects the correct type. The real magic happens in the `BranchFactory` objects, which are returned by the splitters. Basically, the factory assembles the branch node which was picked by `_branch_selector`.

Reply:
To be more specific: the new tree base classes require that all the descendant nodes are passed as parameters when building the branch (decision) nodes. That was not the case previously. We used to rely on the `set_child` method, which was called inside a loop after the (RIP) `SplitNode` was created. Now, we select the branch node type (numerical and binary, numerical and multiway, nominal and binary, or nominal and multiway), build its descendants, select any other needed initialization parameters and pass everything to the `assemble` method of `BranchFactory`.

`BranchFactory` instances are built by the attribute splitters. They carry all the basic information to build a decision node: the split attribute, threshold (or any other needed information), the children's stats, and so on. But the tree splitters "don't know" which kind of decision node the trees might come up with. For example, most of the decision trees rely on the vanilla branch nodes, but EFDT, HAT{C/R} implement some additional fancy logic in their decision nodes. Therefore, the `BranchFactory` instances take everything they can from the tree splitters; the remaining information is obtained from the actual decision trees. The `assemble` call in the `_attempt_to_split` method joins everything and ensures the decision nodes are properly created.

This whole factory thing was what I was able create to solve the problem with the decision nodes. It took me while to figure out. In fact, I spent most of the time during this refactoring creating and polishing the `BranchFactory` class and making sure it could handle every fancy/alien branch nodes I might throw at it.

Reply:
Thanks for the clarification! Just as a comment, we could use `_branch_type_selector` to avoid confusion. Not an actual change request ;-). I leave it up to you.